File: 19. Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
        header = head
        curr = head
        prev = None
        for i in range(n - 1):
            header = header.next
        while header.next is not None:
            header = header.next
            prev = curr
            curr = curr.next
        if prev is None:
            return head.next
        prev.next = curr.next
        return head

File: 19. Remove Nth Node From End of List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode fast = head;
        ListNode slow = head;
        ListNode slowPrev = head;
        for (int i=0; i<n; i++) {
            fast = fast.next;
        }
        if (fast == null) {
            head = head.next;
            return head;
        }
        while (fast != null) {
            fast = fast.next;
            slowPrev = slow;
            slow = slow.next;
        }
        slowPrev.next = slow.next;
        return head;
    }
}